### ISSUE TITLE
Incorrect name & icon reported for media controls in GNOME

### ISSUE THREAD

**Issue by polyjitter** [Mastermindzh/tidal-hifi#511]
Please see as follows:

![image](https://github.com/user-attachments/assets/7b455a5f-556e-4167-9e06-f247049521f3)

I have confirmed this happens across source running, the AppImage, and the Flathub flatpak. It appears at some point in the chain, something is failing to report the right app name and the GNOME media controls don't pull a symbolic icon or a properly formatted name.

I tried patch-desktop-name as a build command in the flatpak manifest, but this did not resolve the issue. It appears to be deeper within the application. Thank you in advance if this could be fixed!

**Comment by polyjitter** [comment]
I think I found the original culprit for the issue:

https://github.com/polyjitter/tidal-hifi/blob/ae699887b2a4ffac4ff8c1f843d64c008030b09b/src/preload.ts#L424-L440

It looks like it's hardcoded to be named "tidal-hifi" in `preload.ts`. Unfortunately, I really don't know enough about Electron to know how to fix this... Is there metadata that could be more easily pulled from to make sure it works no matter what it's called?

To note, it DOES seem to fix the issue when changing the name in `package.json` to `Tidal Hi-Fi`, so my diagnosis might not be completely accurate.

**Comment by Mar0xy** [comment]
According to Electron docs it prefers the productName field in package.json instead of name

```txt
app.getName()

Returns string - The current application's name, which is the name in the application's package.json file.

Usually the name field of package.json is a short lowercase name, according to the npm modules spec.
You should usually also specify a productName field, which is your application's full capitalized name, and which will be preferred over name by Electron.
```

**Comment by Mastermindzh** [comment]
I also remember something about not allowing capitals in mpris 🤔.
Anyway, will have to have a look

**Comment by polyjitter** [comment]
Thank you! Please do note that GNOME has media controls with MPRIS disabled in settings. With it enabled, sometimes it seems to double controls. This leads me to believe there are two media controls that GNOME is picking up (pulled from where or how??? I'm not sure) - just a regular media output without MPRIS enabled, and specifically the MPRIS output with PRIS enabled.

For the regular (I think) GNOME controls, which I'm thinking don't use MPRIS, I think it could be down to two things right now:

* No symbolic icon - I committed a symbolic icon to the Flathub build manifest, not sure how to submit it here or package it for use. I haven't tested that yet, will when I get home from work today. I think the generic icon here is due to that.
* `productName`, as @Mar0xy mentioned. I actually might have been wrong - I might have added a `productName` instead of changing the name, which solved the name issue. Again, I'm going to test that as a change when I'm back on my Fedora install.

For the MPRIS controls, I'm not quite sure unfortunately - I am not too well attuned to how MPRIS works. It's possible it's not worth solving since GNOME has the native controls (still not sure how or why), but it might be worth fixing anyways if other environments this might run in want a product name style formatting.

Also gonna tag @razzeee, who was working a bit on the flatpak manifest to add patch-desktop-name. He was a bit stumped at the time, but if there's any further knowledge he has maybe he can share. That is, if he has the time and wants to, no pressure to respond - he's already dedicated time he never needed to for this small papercut (thank you very much, by the way).

**Comment by razzeee** [comment]
I don't think there are native things other then mpris, at least there shouldn't be. Can you screenshot that setting, I never saw that in gnome?

**Comment by polyjitter** [comment]
Sorry, I should have been more clear. There's an option in Tidal Hi-Fi itself that enables/disabled MPRIS support (maybe someone else can take a screenshot?

With it off, there's only one set of media controls in GNOME. Turning it on (and restarting the application) results in duplicated controls.

It's possible that Electron is already automatically serving an MPRIS player, which is what is causing the duplication when it's enabled. It could also something else unintended in Tidal Hi-Fi, which would require another bug report. Either way, if only MPRIS serves media controls to GNOME, then that means whatever would fix one would fix both, which would make addressing this a bit easier.

**Comment by razzeee** [comment]
Apps can have multiple MPRIS controls, it depends on the ids they set for it. Not sure what tidal does here, but you should be able to have a look with something like https://flathub.org/apps/org.gnome.dspy 

Not sure how helpful that will be for you :)

![image](https://github.com/user-attachments/assets/6ea2925f-7a02-4db6-b1ca-0a612726a650)

**Comment by polyjitter** [comment]
So, with MPRIS off, I am getting media controls, but no MPRIS bus (in session or system) (this is the source build, not flatpak):

![image](https://github.com/user-attachments/assets/d01d26cd-192d-4854-a3fc-c054a509edc7)

Kind of at a loss.

On the plus side, as you can see from the screenshot, setting productName does give a properly formatted name. kudos to @Mar0xy for that tip.

Also, on the flatpak side, the symbolic icon being installed doesn't fix the icon in the media controls (confirmed I was running the version built from the git manifest). Not sure what else to do on that front.

**Comment by razzeee** [comment]
I guess you also get nothing when you search for `mpris`?

And you're saying that's the source build, but that also does not have a working symbolic icon?

**Comment by polyjitter** [comment]
Yes, that's correct - it's the source build, outside of flatpak. 

There's no symbolic icon for source, so that's no surprise - I'm going to work on submitting a PR for that for this repo, I just need to figure out how it's building and from where so I can know how to package that.

What's more confusing is that the flatpak build doesn't have the symbolic icon (which is where I have been able to push the symbolic icon for now). It still shows up like in the root post for this issue.

I'm also thinking the lack of an MPRIS bus from the source build in D-Spy might be a flatpak sandboxing issue. Give me a second to open the Tidal flatpak and play around in it.

**Comment by razzeee** [comment]
I'm wondering, if https://github.com/Mastermindzh/tidal-hifi/blob/f4d4b1a1df651949c2c2626753faf6c8a172bf91/src/preload.ts#L479 runs, even when you don't have the mpris setting enabled - due to player still getting kinda initialized here https://github.com/Mastermindzh/tidal-hifi/blob/f4d4b1a1df651949c2c2626753faf6c8a172bf91/src/preload.ts#L479 which might be an oversight. That instance then also ends up with a different/incomplete config I guess.

So that might be why you end up with two instances of mpris.

**Comment by polyjitter** [comment]
**Double MPRIS Issue:**

Okay, looks like I was wrong (or I had something weird going on moving back and forth between different versions of Tital Hi-Fi. With MPRIS off, no media controls which is fine. With it on, there are media controls. 

(Also, what I was seeing was indeed a Flatpak sandboxing issue, it couldn't see the MPRIS bus because D-Spy doesn't have access to host machine buses. I could fix that but I don't feel like it right now.)

**So no immediate relevant issues with double media controls!** Or, if they are, it's inconsistent behavior which is really kind of outside of the scope of this issue. I did see it at some point, but I think that can be chalked up to an unrelated issue with either Tidal Hi-Fi or GNOME itself. 

Sorry for the confusion. The point is, it shouldn't be relevant to figure that out to solve the app display issue. All MPRIS buses here.

------

**App Display Issue:**

Moving back to the issue at hand, built from the Flatpak manifest (where both patch-desktop-name and symbolic icon have been implemented), the following is still seen:

![image](https://github.com/user-attachments/assets/66d7f54c-9616-4e3e-bb02-a0aa2e058013)

So, something still isn't pulling what it needs to for the symbolic icon or proper application name to be pulled.

When looking at the MPRIS bus, I see what **should** be the working end result. I'm going to take a look at the generated desktop file to see if I can work anything back from there.

**Comment by polyjitter** [comment]
Okay, so:

`/var/home/.../.local/share/flatpak/exports/share/applications/com.mastermindzh.tidal-hifi.desktop`:

```desktop
[Desktop Entry]
Name=TIDAL Hi-Fi
Exec=/usr/bin/flatpak run --branch=master --arch=x86_64 --command=com.mastermindzh.tidal-hifi com.mastermindzh.tidal-hifi
Terminal=false
Type=Application
Icon=com.mastermindzh.tidal-hifi
StartupWMClass=tidal-hifi
Encoding=UTF-8
GenericName=TIDAL Hi-Fi
Comment=Tidal on Electron with widevine(hifi) support
StartupNotify=true
Categories=AudioVideo;
X-PulseAudio-Properties=media.role=music
MimeType=x-scheme-handler/tidal;
X-Desktop-File-Install-Version=0.28
X-Flatpak-RenamedFrom=tidal-hifi.desktop;
X-Flatpak=com.mastermindzh.tidal-hifi
```

![image](https://github.com/user-attachments/assets/9de389ea-4080-411d-bbc2-3e74e34bd8e2)

@razzeee does this look all right to you?

**Comment by razzeee** [comment]
No, I would expect the desktop entry to be `com.mastermindzh.tidal-hifi` - not sure if that would fix it, but that's what I would expect there.

**Comment by polyjitter** [comment]
Okay, so, from what I'm reading, it appears that the desktop filename was patched in the final app.asar (in what, from my messing around with it, seems to be essentially a bundle which also includes the original package.json)?

![image](https://github.com/user-attachments/assets/b9ee1b48-13ae-48b5-98fa-bf895f1aee56)

However, this doesn't rename the desktop file _itself_, it just changes the `desktopName` value.

Within Tidal Hi-Fi, the `desktopEntry` for MPRIS is being hardcoded to `tidal-hifi`:

https://github.com/Mastermindzh/tidal-hifi/blob/f4d4b1a1df651949c2c2626753faf6c8a172bf91/src/preload.ts#L439C1-L439C36

This probably explains the behavior I'm seeing in both source build and flatpak:

* Source build does not have a desktop file at all yet, so it's defaulting back to tidal-hifi, no icon.
* Flatpak build desktop filename obviously does not match `tidal-hifi`, so MPRIS isn't reading the generated desktop file, defaulting back to tidal-hifi, no icon.

My **hypothesis** is that on a conventional packaged-based install (deb, rpm, etc.), it does indeed at least read the right name (Tidal Hi-Fi). I am on an ostree-based distro right now so I can't easily test this.

As for a solution, I think the `desktopEntry` in `preload.ts` should be pulled from a `desktopName` set in `package.json`. This doesn't exist by default, so maybe it should be added in source build. From there, I would hypothesize things to behave as expected.

Let me know if that all makes sense!

**Comment by polyjitter** [comment]
**YES!** I was right.

I didn't install the RPM, but with a workaround, I was able to confirm that was indeed the issue:

1. run `cp .local/share/flatpak/exports/share/applications/com.mastermindzh.tidal-hifi.desktop ~/.local/share/applications/tidal-hifi.desktop`.
2. Inside the new copy of the desktop file, set `NoDisplay=true`.
3. The symbolic icon and formatted name appear:

![image](https://github.com/user-attachments/assets/a5a01645-c94c-4b2f-a6fa-64ca76d48aa6)

This means that tidal-hifi should probably pull the desktop file name programmatically from package.json. It's not my project, so I don't know how you'd wanna go about that, but that would indeed be the solution.

**Comment by Mastermindzh** [comment]
Hey,

When we renamed `tidal-hifi` to `Tidal Hi-Fi` I tried giving mpris the name `Tidal Hi-Fi` (from the package json even...) but it would fail to register in a lot of cases.

Even though the [dbus-specification](https://dbus.freedesktop.org/doc/dbus-specification.html#message-protocol-names-bus) mentions you can use all kinds of casing.

We could try again of course :)

**Comment by polyjitter** [comment]
Sorry, this is a late reply, but yeah it's possible that it wouldn't register in other environments.

For GNOME specifically, it appears to pull app name and symbolic icon by matching the given name from dbus to a matching desktop file; otherwise, it'll fallback to the given dbus name.

For appimage, etc. builds there is no desktop file, and for flatpak, the desktop file is named something completely different, so it falls back to the reported dbus name instead (tidal-hifi).

So a solution to this issue would be to get it to report to dbus the desktop file name, and if there isn't a desktop file, report either tidal-hifi or Tidal Hi-Fi (depending on which one works). 

Importing package.json in an electron application can lead to significant security risks, though, so not sure the solution here...
